### ISSUE TITLE
Clean up the branches, issues ?, project

### ISSUE THREAD

**Issue by Shagequi** [nextcloud/android#4876]
Hi,

I asked in this PR : https://github.com/nextcloud/android/pull/4865 if it was possible to clean the project.

Here are some ideas to enhance the repo order :
- Removing the stale/no longer used branches
- Create a develop branch and stop using master to develop on (master would become stable)
- Create specific releases branches (which are stables by default) 
- Use tags for each version in master (and create releases, based on tags, and keep releases to allow rollback between stable versions)
(like on IOS project, respect a gitflow process)

About issues, here is a proposal :
If issue is a bug, and is a reproductible bug, keep issue opened while the bug is unfixed/unmerged (in a hotfix or in develop).
But if it's an enhancement, why not create a "TODO" element in Projects and close the issue once approved (or disapproved)  ? 2/3 of issues are asking for enhancements, we could simply use the issues to talk, accept/refuse and describe ideas, but use the Projects part to facilitate the development.

@AndyScherzinger @tobiasKaminsky @jancborchardt What do you think about that ?

**Comment by AndyScherzinger** [comment]
This repo works a bit different (for good reasons amho)

* we use the github flow
* we do have stable branches (feature release stabilization and bugfix releases for a feature release)
* master is always latest dev cut (gitflow renders the master useless in my opinion)
* each version is alreadya tagges see https://github.com/nextcloud/android/tags (for daily, beta, rc, stable)

This works extremely well for us, also running backport bots for backporting fixes from merged PRs to stable branches.

Did you check https://github.com/nextcloud/android/blob/master/CONTRIBUTING.md ?

As for issues:
* If a bug is reproducable it gats labeled as such (and stale bot won't close it)
* Issues get referenced by PRs and auto closed upon merging (Bugs and Features)
* PRs and issues get assigned to milestones (read: releases)
* GH Projects aren't used

So that is my PoV (while I am aware that it rejects nearly all suggestions), I am fine with the stale branch discussion while most stale branches are still valid since we did a housekeepinjg before and removed all obsolete branches.

The above is just my opinion though and should also shed some light on how we are currently operating. (including all the automation).

**Comment by Shagequi** [comment]
I checked Contributing (and watch some of your speechs) @AndyScherzinger but the schema is unfortunately not comparable with reality.
I think that any environment that you understand yourself is comfortable and unworthy of change. The problem is that the branches go in many directions, and I never understand how to work properly.

![Capture d’écran 2019-11-22 à 12 59 09](https://user-images.githubusercontent.com/7050479/69424049-e7ad4780-0d27-11ea-9d11-5efdc5ddf7c8.png)

Just watch the above screenshot, I don't really understand the "flow", if you watch this one :

![Capture d’écran 2019-11-22 à 13 03 37](https://user-images.githubusercontent.com/7050479/69424292-889c0280-0d28-11ea-9828-b4abd3656874.png)

It's really more understandable. And it respects gitflow conventions.

Concerning master, this branch is currently unstable and used as a base for all, with gitflow, the develop branch would replace master, and master would became the stable base for releases, and you create all your tags and all your releases from master (as you do now). 

About issues, I consider these ones as a "mailbox" which allows to communicate with customers, developers or project managers. But, once a letter has been read, and considered as "to be processed", we don't left in the mailbox, but we put it in "processing bins" specially designed for this purpose, Project on GH are "processing bins".

I am open to criticism and discussion, in my opinion sharing our feelings is extremely beneficial to the project, why do you think that this project works better in this way than in an other using gitflow or my proprosals ?

**Comment by tobiasKaminsky** [comment]
Hey,

thank you very much for your interest in helping and making the process better.

So here are my thoughts
I do not see any benefit in having another develop branch next to master. We currently develop on master which is unstable. Once we want to release a new version (e.g. 3.10) we create a new stable-3.10 branch. If bugs are found, we fix them on master and backport them to stable-3.10.
If we, meanwhile, need to create a hofix release for 3.9, we can simply do this in stable-3.9.
--> what would be the benefit of having "develop" branch?

Having the issues with enhancement requests open is a project wide decision, so that needs to stay consistent. Server e.g. also has ~800 open issues with enhancement.

Projects: we do not have a real roadmap plan for each release. The idea is that each contributor can pick up any enhancement they like, discuss it, create a PR, and when it is ready, it is merged. 
We do not wait for any PR to be finished until we do a release, but rather do it the other way around: what is in, will be released.
However there are from time to time features that should be released approx. in sync with the next server release, but still this is not set in stone.
Also there are some "overview" issues, which describe a broader roadmap/ way what we want to do in mid/long term.

Regarding branch name: seeing your picture with your branches, it seems you follow the idea like in server, with having categories in branch names, e.g:
- enh/123/addNewFeatures
- bug/123/fixNPEinDetail
is this correct?
Can you describe this a bit more, and have a look if this is also the same as in server repo?

All in all nothing is set into stone, as it is also only my personal opinion.
I hope you still want to contribute to this project and if you have specific questions regarding roadmap or other related questions, feel free to open up a new issue.

**Comment by Shagequi** [comment]
@tobiasKaminsky 

Hi,

No repo can achieve perfection, what I propose simply harmonizes your work, and the work of the people who want to contribute with you.

Having like 30 branches in "root" (at the same level than master) it's in my opinion, inconsistent, you can't work properly; and name his branch according to his nickname is ... special. 

Start by just having master, stable branches (named stable/3.9 instead of stable-3.9), features and backports would allow you to have an already much cleaner project (while keeping your way of working).

And here is my vision of gitflow : 

![image](https://user-images.githubusercontent.com/7050479/69438294-10900580-0d45-11ea-826d-35a38e05fcb2.png)

So it "could be correct" yes, by using the `/` to separate branches between "folders", but I think it's important to keep a clear root/base.

Concerning issues, I respect and understand your vision, although I don't fully approve it :)

**Comment by ezaquarii** [comment]
@Shagequi what kind of **problem** are you trying to solve? I fail to see it.

> and name his branch according to his nickname is ... special.

For all the practical purposes it works the same way as a repository fork, with minor exception that all discussions stay in Nextcloud organization. It's very reliable, lightweight and intuitive form of namespacing. Those branches are deleted after merge anyway.

If you are confused how to work with Nextcloud, this is how it works for external contributors such as you and me:

0. ignore everything
1. branch from `master`
2. just do your job
3. tag it as `Review`
4. @tobiasKaminsky will come, review and eventually merge it. :)

Stale branches can become a concern at some point, but I think we haven't got there yet.

**Comment by ezaquarii** [comment]
To remove stale branches from local repo you can use this command:
```
git prune origin
```

**Comment by jancborchardt** [comment]
First let’s keep in mind that this issue contains a lot of different things. Naturally some we agree on, some are already done like this (or almost the same) and some we disagree on. This doesn’t mean everything is bad, and of course as always some things can be improved here and there. :heart: 

--------------------------

Regarding stale branches, it’s just 8 at the moment, all with pull requests open: https://github.com/nextcloud/android/branches/stale
To help cleaning those up, it’s best to review the pull requests and give feedback, or help where needed. :)

-------------------------------

Regarding branch conventions, across Nextcloud we keep stuff simple and do _not_ use Git Flow, but Git**Hub** Flow which is much simpler:
- Understanding the GitHub flow (short intro): https://guides.github.com/introduction/flow/
- Issues with git-flow: https://githubflow.github.io/
- Git Flow vs Github Flow: https://lucamezzalira.com/2014/03/10/git-flow-vs-github-flow/

It’s still probably indeed helpful to prefix branches with `bug/<number>/branch-title` or `enhancement/…` like we do in server, so good feedback there.


-------------------

The rest was already commented on nicely by @AndyScherzinger @tobiasKaminsky @ezaquarii :)

**Comment by AndyScherzinger** [comment]
> It’s still probably indeed helpful to prefix branches with bug/<number>/branch-title or enhancement/… like we do in server, so good feedback there.

This is an issue that became clear to me with @Shagequi's comment and especially if this is alrewady the case for the server repo (which has the most contributors) than it makes even more sense (for consistency reasons) to adapt the naming schema.

Other than that should we raise this with the iOS repo, looping in @marinofaggiana for feedback

**Comment by Shagequi** [comment]
About naming process, it's a first thing which is necessary, I believe.

![Capture d’écran 2019-11-26 à 11 02 41](https://user-images.githubusercontent.com/7050479/69619732-4fbba080-103c-11ea-91f4-e89e36eb7c93.png)

After having pruned the remote branches, we arrive at 22 branches in "root" vs 39 before prune.

I find that 22 branches are not really precise, and in example, we don't know what the "glide" branch means (do we replace picasso with glide? Do we remove glide?), we don't know what it means "fix-2018" (Is this all fixes in 2018? If we want to fix something now we have to create a branch fix-2019 ? Or create `fix/name-of-fix` ?) 
Github Flow is one thing for people to be able to contribute without having to worry about it, and it's adapted for a small project, or a project with few people involved. But for the immensity of this project, and its complexity, I thought GitFlow would be more suitable. (If you think that gitflow is not needed/adapted, let us at least do a little tidying up to call for adequate tidying up for future contributors ? By organizing work in prefixed branches (`/`) in order to reduce their amount, then to ask people to set branch names a little bit explicit)

In addition, you are extremely open to changes from people on all sides, which means that a good arrangement of branches and committees would be beneficial for you (code reviewers) and for us (volunteers who develop stuff).

Thank you for your reading ❤️

**Comment by jancborchardt** [comment]
> After having pruned the remote branches, we arrive at 22 branches in "root" vs 39 before prune.

Right – and this is unfortunately how git does it. This has nothing to do with how the repository is organized.

You can see from the info on the top of https://github.com/nextcloud/android/  that there are only 24 branches. Of course fluctuating: new ones appear with new pull requests, and when pull requests are merged the branch is automatically deleted.


> I find that 22 branches are not really precise, and in example, we don't know what the "glide" branch means (do we replace picasso with glide? Do we remove glide?), we don't know what it means "fix-2018" (Is this all fixes in 2018? If we want to fix something now we have to create a branch fix-2019 ? Or create `fix/name-of-fix` ?)

This is why I said above:
Regarding stale branches, it’s just 8 at the moment, all with pull requests open: https://github.com/nextcloud/android/branches/stale
To help cleaning those up, it’s best to review the pull requests and give feedback, or help where needed. :)

Then we would be down to 16 branches already, which are all relevant pull requests so absolutely have reason to exist. :) 


> Github Flow is one thing for people to be able to contribute without having to worry about it, and it's adapted for a small project, or a project with few people involved. But for the immensity of this project, and its complexity, I thought GitFlow would be more suitable.

Github Flow is absolutely fine. We use it for the server repository since ages, which has many more contributors:
https://github.com/nextcloud/server/
No reason to overcomplicate it.


> (If you think that gitflow is not needed/adapted, let us at least do a little tidying up to call for adequate tidying up for future contributors ? By organizing work in prefixed branches (`/`) in order to reduce their amount, then to ask people to set branch names a little bit explicit)

Yes, again, agreed on this. :) And @AndyScherzinger seems to agree too. I’m sure @tobiasKaminsky @ezaquarii and others find this a good enhancement as well?

**Comment by AndyScherzinger** [comment]
> And @AndyScherzinger seems to agree too.

I do, while I think we still need to define/agree on the naming pattern since the server doesn't do this consistently (see `/fix` and `/bugfix` or `/enh` and `enhancement` and `feature`).

**Comment by ezaquarii** [comment]
TL;DR Whatever works for @tobiasKaminsky - he's the only one working with many branches at once.

@jancborchardt @AndyScherzinger I still don't understand what kind of **problem** we are trying to solve here?

I think that workflow should be choosen by the people responsible for integration, as they need to naviagte all of it. This person is @tobiasKaminsky . It's hard to theorieticize about his pain points while not being in his shoes and he was pretty clear that he's happy with his current way of doing things. It doesn't matter what people like, dislike or use in their workplace - it is @tobiasKaminsky who does the integration **here**.

Everybody else just care about 2 labels really: master and whatever they work on this week and will be most likely ignoring everything else.

Whatever you adopt please bear in mind that every new contributor - and many of them are one-time visitors - must be educated about the process and apply it. Current workflow does not need any additional effort, as it is the default mode everybody familiar with Git will try first.

I'll refrain myself from publishing my opinion about Git Flow itself, but here is my factual observation:

I don't see Git Flow being popular. Every single company I worked in had this discussion and I never seen it adopted. It got plenty of social media traction for sure, because I was nagged with a link to the marble diagram many, many times back in the days when people were moving from centralized VCS-es, trying to  re-use their knowledge of centralized workflows (it was 2010), but today - not so much really, I simply don't see it. While I'll stay silent about my feelings about GF, please search for "gitflow good or bad" in google and see what people have to say about it.

If the advocacy for gitflow is that people are using it - I don't see numerical evidence supporting this idea.

**Comment by AndyScherzinger** [comment]
> TL;DR Whatever works for @tobiasKaminsky - he's the only one working with many branches at once.

I can agree with this while consistency with the server repo would be good too.

> I still don't understand what kind of **problem** we are trying to solve here?

None, we try to **prevent confusions** and for me this *just* means that we adopt a certain naming patter *within* this repo (not for fork repos)

> Whatever you adopt please bear in mind that every new contributor - and many of them are one-time visitors - must be educated about the process and apply it.

Nope, this would just apply for branches created within this repo, not in a fork. So just orga members with push rights.

> I'll refrain myself from publishing my opinion about Git Flow itself

So do I (same opinion I guess)

**Comment by tobiasKaminsky** [comment]
We can have a different branch naming, sure, but then let us have the same nextcloud wide and encourage all people to use it.



> I find that 22 branches are not really precise, and in example, we don't know what the "glide" branch means (do we replace picasso with glide? Do we remove glide?), we don't know what it means "fix-2018" (Is this all fixes in 2018? If we want to fix something now we have to create a branch fix-2019 ? Or create `fix/name-of-fix` ?)

I know this is just an example, but to show, that I (and maybe other do not work this way):
I do not care how branch "glide" is named. I just go to github, see a pr to review, c&p branch name, check it out, etc.
So, while new branching scheme will help if you see only branch names, it does not really have an impact/benefit (for me) on my daily work.

**Comment by jancborchardt** [comment]
> > And @AndyScherzinger seems to agree too.
> 
> I do, while I think we still need to define/agree on the naming pattern since the server doesn't do this consistently (see `/fix` and `/bugfix` or `/enh` and `/enhancement` and `/feature`).

So if people want to use a possibly improved naming scheme, it should follow the relevant GitHub labels, which are `bug` or `enhancement`.

If not that doesn’t hurt either. It’s totally fine if some branches are named with the prefix and some not, change also doesn’t happen immediately.

And regarding the amounts of branches, again it’s best if people help review the older pull requests at https://github.com/nextcloud/android/branches/stale

------------------------


I feel that this discussion is done, right?

- Nextcloud uses Git_Hub_ flow cause it’s simple and works even for huge projects.
- If people want to use `bug/` or `enhancement/` as branch prefixes, no one stops them.
- Help review the older pull requests at https://github.com/nextcloud/android/branches/stale
- For any issue, it makes sense to open _separate_ issues about _separate_ concerns.

Can we close this?

**Comment by tobiasKaminsky** [comment]
> * Help review the older pull requests at https://github.com/nextcloud/android/branches/stale

--> not all pull requests are ready for review. Some are in developing phase and "only" in repo for transparency reasons or to get feedback as early as possible.

**Comment by Shagequi** [comment]
Thank you for these answers.
We may close this issue now.
